File: 215. Kth Largest Element in an Array/main.go
package main

import "math/rand"

func findKthLargest(nums []int, k int) int {
	return QSearch(nums, k)
}

func QSearch(arr []int, k int) int {
	shuffle(&arr)

	return sort(&arr, 0, len(arr)-1, len(arr)-k)
}

func sort(nums *[]int, start int, end int, pos int) int {
	pivotPos := partition(nums, start, end)

	if pos == pivotPos {
		return (*nums)[pivotPos]
	} else if pos > pivotPos {
		return sort(nums, pivotPos+1, end, pos)
	} else {
		return sort(nums, start, pivotPos-1, pos)
	}
}

func partition(arr *[]int, start int, end int) int {
	pivot := (*arr)[start]

	i, j := start+1, end
	for i <= j {
		for i < end && (*arr)[i] <= pivot {
			i++
		}

		for j > start && (*arr)[j] > pivot {
			j--
		}

		if i >= j {
			break
		}
		swap(arr, i, j)
	}

	swap(arr, start, j)

	return j
}

func shuffle(arr *[]int) {
	n := len(*arr)
	for i := 0; i < n; i++ {
		r := i + rand.Intn(n-i)
		swap(arr, i, r)
	}
}

func swap(arr *[]int, a int, b int) {
	temp := (*arr)[a]
	(*arr)[a] = (*arr)[b]
	(*arr)[b] = temp
}

File: 215. Kth Largest Element in an Array/main.py
import random
from typing import List


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        shuffle(nums)
        lo, hi = 0, len(nums) - 1
        ps = len(nums) - k
        while lo <= hi:
            p = partition(nums, lo, hi)
            if p < ps:
                lo = p + 1
            elif p > ps:
                hi = p - 1
            else:
                return nums[p]
        partition(nums)


def partition(arr, lo, hi):
    pivot = arr[lo]
    i = lo+1
    j = hi
    while i <= j:
        while i <= hi and arr[i] <= pivot:
            i += 1
        while j > lo and arr[j] > pivot:
            j -= 1

        if i >= j:
            break

        swap(arr, i, j)
    swap(arr, j, lo)
    return j

def shuffle(arr):
    for i in range(0,len(arr)):
        rand = random.randint(0, len(arr)-1)
        swap(arr, i, rand)

def swap(arr, a, b):
    arr[a], arr[b] = arr[b], arr[a]

s = Solution()
s.findKthLargest([2,1], 1)